unified-runtime/source/adapters/level_zero/async_alloc.cpp

```diff
@@ -1,43 +1,263 @@
-//===--------- async_alloc.cpp - CUDA Adapter -----------------------------===//
+//===--------- async_alloc.cpp - Level Zero Adapter -----------------------===//
 //
-// Copyright (C) 2024 Intel Corporation
+// Copyright (C) 2025 Intel Corporation
 //
 // Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
 // Exceptions. See LICENSE.TXT
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
 //===----------------------------------------------------------------------===//
 
+#include "context.hpp"
+#include "enqueued_pool.hpp"
+#include "event.hpp"
+
+#include "logger/ur_logger.hpp"
+
+#include <umf_helpers.hpp>
 #include <ur_api.h>
 
 namespace ur::level_zero {
 
-UR_APIEXPORT ur_result_t urEnqueueUSMDeviceAllocExp(
-    ur_queue_handle_t, ur_usm_pool_handle_t, const size_t,
-    const ur_exp_async_usm_alloc_properties_t *, uint32_t,
-    const ur_event_handle_t *, void **, ur_event_handle_t *) {
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+static ur_result_t enqueueUSMAllocHelper(
+    ur_queue_handle_t Queue, ur_usm_pool_handle_t Pool, const size_t Size,
+    const ur_exp_async_usm_alloc_properties_t *Properties,
+    uint32_t NumEventsInWaitList, const ur_event_handle_t *EventWaitList,
+    void **RetMem, ur_event_handle_t *OutEvent, ur_usm_type_t Type) {
+  std::ignore = Pool;
+  std::ignore = Properties;
+
+  std::scoped_lock<ur_shared_mutex> lock(Queue->Mutex);
+
+  // Allocate USM memory
+  ur_usm_pool_handle_t USMPool = nullptr;
+  if (Pool) {
+    USMPool = Pool;
+  } else {
+    USMPool = &Queue->Context->AsyncPool;
+  }
+
+  auto Device = (Type == UR_USM_TYPE_HOST) ? nullptr : Queue->Device;
+
+  std::vector<ur_event_handle_t> ExtEventWaitList;
+  ur_event_handle_t OriginAllocEvent = nullptr;
+  auto AsyncAlloc =
+      USMPool->allocateEnqueued(Queue, Device, nullptr, Type, Size);
+  if (!AsyncAlloc) {
```

Comment:
Won't this bypass any limits specified for async pools?

Reply:
I would say that it's possible since we are caching memory from those pools without really releasing it back to the pool immediately. The implementation would probably need to introduce some restriction for caching based on the pool limits.